Feat: toggle time/date on SD watchface

bradmartin opened the pull request.

bradmartin commented:
Working on a little polish to the watchface layout. Currently, when you toggle the time/date there is a subtle shift in the layout. This is caused by the XML layout and some View attributes and how we handle the `Visibility` of the Views/TableRows during the event. Have an idea on some fixes, just takes time to test the various routes.

bradmartin commented:
Related #820

bradmartin commented:
Should be ready for more testing on actual hardware. Looking much better across 4 different emulators with different screen sizes.

finger563 commented:
Sounds good - i'll pull and test on the pro 3 and c2+

bradmartin commented:
@finger563 this seemed like the best way to handle sizing that image - https://github.com/Max-Mobility/permobil-client/pull/832/files#diff-4a5c40d51cc6d51b6af627b82774d149R326 bc we already scale it with `scaleType` attribute. It looks good on all the emulators I have running. We can adjust that number a bit as needed of course.

finger563 commented:
@bradmartin  does this handle the pages that show up when you go to configure the watchface as well? i've noticed (in the past) that the layouts weren't necessarily identical

bradmartin commented:
the page that shows when you go to configure the watchface is actually a `PNG`  inside the `res/drawable` titled `digital_watchface_preview.png` it's specifically a `nodpi` image which is supposed to scale accordingly. 

So we'll need to modify the image for the watchface preview with any changes to make it 1:1

bradmartin commented:
@finger563  - can you pull this and run it on the C2 and Pro3? I know I've corrected a few of the items on the E2 and emulators.

One thing that is killing me is when you are in the "date mode" and enter ambient, the time doesn't show... even though the TableRow is set equal to VISIBLE. I've stepped thru the debugger and it all looks correct, so I'm currently looking at that. 

I've also reverted back the font sizes so they should be what they used to be.

The button should be centered vertically now as well. It looks good on the E2 with being centered.

finger563 commented:
Sure - will check shortly 👍

bradmartin commented on apps/wear/smartdrive/app/App_Resources/Android/src/main/java/com/permobil/smartdrive/wearos/watchface/digital/DigitalWatchFaceService.java:
```diff
@@ -648,8 +694,42 @@ private void openSmartDriveAppOrPlayStore() {
+            // if we are in ambient mode both TIME and DATE need to show
+            // else if we are in TIME TEXT mode, show the time and hide the date
+            // else hide the date
+            if (isInAmbientMode()) {
```
This is where the issue was with showing the time when we would enter ambient and time would not render. Logic tests out, but quick look over wouldn't hurt @finger563, my brain hurts after tracking this one down and trying to ensure it was correct 🙃

bradmartin commented:
https://github.com/Max-Mobility/permobil-client/pull/832/files#r475887406 @finger563

finger563 commented on apps/wear/smartdrive/app/App_Resources/Android/src/main/java/com/permobil/smartdrive/wearos/watchface/digital/DigitalWatchFaceService.java:
```diff
@@ -648,8 +694,42 @@ private void openSmartDriveAppOrPlayStore() {
+                // show the date and hide the time
+                dateTableRow.setVisibility(View.VISIBLE);
+                timeTableRow.setVisibility(View.GONE);
+                return;
```
Why the early return here if no in ambient mode? do we not need to call the `colorTextViewsForAmbientHandling()` down below?

bradmartin replied:
I don't think we need to, as the coloring happens BEFORE this during a different event. I believe it gets handled in the `onAmbientChanged` so I was thinking if it's just the `date` value, we don't need to worry about all the calculations for the time text value, so just return out of it. Testing and it seems to work as expected, so I think we're good but obviously if you notice odd behavior with the coloring then we'd just drop the `return`.

finger563 commented on apps/wear/smartdrive/app/App_Resources/Android/src/main/java/com/permobil/smartdrive/wearos/watchface/digital/DigitalWatchFaceService.java:
```diff
@@ -742,18 +808,18 @@ private float getSmartDriveBatteryLevel() {
 
         private void colorTextViewsForAmbientHandling() {
             if (!isInAmbientMode()) {
-                hourTextView.setTextColor(whiteColor);
-                colonTextView.setTextColor(whiteColor);
-                minuteTextView.setTextColor(whiteColor);
-                amPmTextView.setTextColor(whiteColor);
-                dateTextView.setTextColor(whiteColor);
+                hourTextView.setTextColor(-1);
+                colonTextView.setTextColor(-1);
+                minuteTextView.setTextColor(-1);
+                amPmTextView.setTextColor(-1);
+                dateTextView.setTextColor(-1);
             } else {
                 // in ambient so use ambient color
-                hourTextView.setTextColor(ambientColor);
-                colonTextView.setTextColor(ambientColor);
-                minuteTextView.setTextColor(ambientColor);
-                amPmTextView.setTextColor(ambientColor);
-                dateTextView.setTextColor(ambientColor);
+                hourTextView.setTextColor(-1);
+                colonTextView.setTextColor(-1);
+                minuteTextView.setTextColor(-1);
+                amPmTextView.setTextColor(-1);
+                dateTextView.setTextColor(-1);
```
I'm confused by these - why the change from `whiteColor` / `ambientColor` to `-1` - and why bother with the conditional if we're setting it to the same value in both places?

bradmartin replied:
I was just cleaning up, dropping the `getColor` methods bc the values we have for those specific colors we can use the `int` values so just using those instead of the variables. And yea, if we set it to the same, then we can move it from the conditional 👍 didn't notice that when running thru it.

bradmartin replied:
I believe the reason they both resolve to -1 is because they're both basically `Color.WHITE` so the system was always returning `-1` for both of those variables when we would call into the THEME and use `getColor()` on the colors.xml files 👍  so by going this route I dropped a bit of code and several operations that were occurring during the `onDraw()` which fires every second in active mode, so just seemed it might be more efficient if we know the system `int` to use, but yea we can even drop the conditional in that case.

finger563 commented:
Yea so far the testing looks good - coloring is good and positioning looks good as well. The size / emphasis on the time text in ambient mode is still a lot smaller than it used to be, we should probably bump that back up so it's an easier glance when in ambient mode

bradmartin commented:
I'll look back and compare the previous ambient sizes, I thought I had reverted it. I do think the SD app we use a much higher font size in ambient so I'll try to see if we can't 1:1 that but it might be a little different depending how the size is set on the NS `View` but I'll look and get this cleaned up and I think we'll be okay for now.

bradmartin commented:
Font size fixed during ambient change. Took bit to track down bc I had set the `scaling` of the font in the XML layout to "uniform" which was an effort to try and resize the date if it "overflowed" the first line to avoid hitting the circles in the layout originally. Having that set, caused the actual code `setTextSize()` to not do anything... so yeah head scratcher digging thru the previous changes.

@finger563 so this PR is good unless we still need to update the config layout with any changes, I'm gonna move to other items in the meantime, while you can confirm that and the font size changes working well on other watches.

finger563 commented:
Yea I wondered about that w.r.t. the lag on tapping - it's definitely not perfect now, but it is functional (the tapping).  The sizing looks good on both the C2+ and the Pro 3, so I think we're good to go on updating the layout xml for the configuration page.
